src/backend/services/metrics.py

```diff
@@ -0,0 +1,161 @@
+import asyncio
+import hashlib
+import logging
+import os
+import threading
+import uuid
+
+from httpx import AsyncHTTPTransport
+from httpx._client import AsyncClient
+from starlette.middleware.base import BaseHTTPMiddleware
+from starlette.requests import Request
+from starlette.responses import StreamingResponse
+
+from backend.chat.collate import to_dict
+from backend.schemas.metrics import MetricsData
+
+REPORT_ENDPOINT = os.getenv("REPORT_ENDPOINT", "")
+
+import time
+
+from starlette.responses import Response
+
+
+class MetricsMiddleware(BaseHTTPMiddleware):
+    async def dispatch(self, request: Request, call_next):
+        request.state.trace_id = str(uuid.uuid4())
+        request.state.agent = None
+
+        start_time = time.perf_counter()
+        response = await call_next(request)
+        duration = time.perf_counter() - start_time
+
+        data = self.get_data(request.scope, response, request, duration)
+        threading.Thread(target=report_metrics_thread, args=(data,)).start()
+        return response
+
+    def get_data(self, scope, response, request, duration):
+        data = {}
+        trace_id = None
+        if hasattr(request.state, "trace_id"):
+            trace_id = request.state.trace_id
+
+        if scope["type"] != "http":
+            return None
+
+        data = MetricsData(
+            method=self.get_method(scope),
+            endpoint_name=self.get_endpoint_name(scope, request),
+            user_id_hash=self.get_user_id_hash(request),
+            success=self.get_success(response),
+            trace_id=trace_id,
+            status_code=self.get_status_code(response),
+            object_ids=self.get_object_ids(request),
+            agent=self.get_agent(request),
+            duration=duration,
+        )
+
+        return data
+
+    def get_method(self, scope):
+        try:
+            return scope["method"].lower()
+        except KeyError:
+            return "unknown"
+        except Exception as e:
+            logging.warning("Failed to get method:", e)
```

Comment:
nit, we use fstrings everywhere else to format our warning logs. lets do the same here?